### PR TITLE
Fix huge message striping

### PR DESCRIPTION
**Pull request by wesbland**
## Pull Request Description

90a02540728c broke huge message striping because it started filling in the current offset earlier than was previously expected. This meant that the control message describing the message hadn't yet arrived.

Change the huge message receive protocol to use both the current offset and the message size to make sure the message has arrived.

Also, the initialization for the communicator hints failed to take into account the CVARs that the user can set to apply to the entire job.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by wesbland** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
> [90a0254](https://github.com/pmodels/mpich/commit/90a02540728c9ce30d990312740f538d0883c15b) broke huge message striping because it started filling in the current offset

90a0254 broke huge message striping because it started filling in the current offset
Was this the culprit? Looks like an accident.

**Comment by wesbland** [comment]
> > [90a0254](https://github.com/pmodels/mpich/commit/90a02540728c9ce30d990312740f538d0883c15b)
> > broke huge message striping because it started filling in the current offset
> 
> image
> Was this the culprit? Looks like an accident.

Yep. This was what did it. I wasn't sure what the need for this change was, but adding the fix here isn't a big deal.

**Comment by wesbland** [comment]
Let me clarify. That code wasn't removed, it was just moved somewhere else.

**Comment by raffenet** [comment]
> Let me clarify. That code wasn't removed, it was just moved somewhere else.

The issue was that I didn't have access to the comm_ptr, so I couldn't get the striping information. I thought that I would have it later when the copy actually started, but I must have missed an edge case.

**Comment by wesbland** [comment]
The tests all passed in the last run. I rebased this, but don't plan to run the tests again. @raffenet / @hzhou feel free to merge when you think this is good.

File: src/mpid/ch4/netmod/ofi/ofi_events.c [hunk 1]
```diff
@@ -321,9 +321,9 @@ int MPIDI_OFI_get_huge_event(struct fi_cq_tagged_entry *wc, MPIR_Request * req)
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_GET_HUGE_EVENT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_GET_HUGE_EVENT);
 
-    if (recv_elem->localreq && recv_elem->cur_offset != 0) {    /* If this is true, then the message has a posted
-                                                                 * receive already and we'll be able to find the
-                                                                 * struct describing the transfer. */
+    /* If this is true, then the message has a posted receive already and we'll be able to find the
+     * struct describing the transfer. */
+    if (recv_elem->localreq && recv_elem->cur_offset != 0 && recv_elem->remote_info.msgsize != 0) {
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_events.c:326]: Do we understand when `localreq` will be set or unset, when `cur_offset` will be `0` or non-zero, and `when `remote_info.msgsize` will be `0` or non-zero?

Right now reading this code, we are looking at (2*2*2 = 8) possibilities and it requires tracing a lot of code to understand. If we can narrow down the condition and either refactor the condition or document the conditions (causes of the conditions), that will help the maintainence.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: > Do we understand when `localreq` will be set or unset, 

When the message is posted. 

> when `cur_offset` will be `0` or non-zero, 

When the first part of the message has arrived. 

> and `when `remote_info.msgsize` will be `0` or non-zero?

Confirming that we didn't underflow. We might be able to drop this part now that we're adding the previous condition.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: This isn't about handling lots of cases. We only want to handle one and drop all of the rest.

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: Thanks. That is helpful. I guess my real question is, what are the cases that we drop?

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: All the cases where a message hasn't been matched yet (e.g, unexpected only or posted but not received).

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: Alright, I spent my time understanding the logic, cross-referencing all the relevant functions. My understanding results in #5364. 

When reading `MPIDI_OFI_get_huge_event`, the natural question (as I asked) is when this function is being called (its roles/purposes), and if this function only processes some cases, what happens to the other cases. Both questions can't be understood without reading all other code and put all the code together.

Now, if we move the condition to the caller site, the context makes it really simple. When recv is received, we check whether control message has been recved. If yes, start `MPIDI_OFI_get_huge_event`; otherwise, leave for the control message handler to do that.

When the control message handler is called, we check whether receive has been posted. If yes, start `MPIDI_OFI_get_huge_event`; otherwise, leave for the recv to do that.

Both logic is straightforward and can be understood by just understand locally. In particular, I don't need read a single line in `MPIDI_OFI_get_huge_event` to understand it.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_events.c:326]: It was clear enough to me before, but if you prefer to refactor things, I won't stop you...

**Comment by wesbland** [comment]
There was one more commit in this PR, but I'll just leave that in #5301 and close this.
